Error occurred while enabling AntiRedstoneClock-Remastered v2.0.13

### Server Logs

[latest.log](https://github.com/user-attachments/files/22184378/latest.log)

### Expected behavior

Well, the plugin run

### Actual behavior

[09:44:21 INFO]: [AntiRedstoneClock-Remastered] Enabling AntiRedstoneClock-Remastered v2.0.13
[09:44:21 ERROR]: Error occurred while enabling AntiRedstoneClock-Remastered v2.0.13 (Is it up to date?)
java.lang.NoClassDefFoundError: net/kyori/adventure/text/minimessage/translation/MiniMessageTranslationStore
        at AntiRedstoneClock-Remastered-2.0.13.jar/net.onelitefeather.antiredstoneclockremastered.AntiRedstoneClockRemastered.onEnable(AntiRedstoneClockRemastered.java:70) ~[AntiRedstoneClock-Remastered-2.0.13.jar:?]
        at org.bukkit.plugin.java.JavaPlugin.setEnabled(JavaPlugin.java:288) ~[paper-api-1.20.6-R0.1-SNAPSHOT.jar:?]
        at io.papermc.paper.plugin.manager.PaperPluginInstanceManager.enablePlugin(PaperPluginInstanceManager.java:202) ~[paper-1.20.6.jar:1.20.6-151-a4f0f5c]
        at io.papermc.paper.plugin.manager.PaperPluginManagerImpl.enablePlugin(PaperPluginManagerImpl.java:109) ~[paper-1.20.6.jar:1.20.6-151-a4f0f5c]
        at org.bukkit.plugin.SimplePluginManager.enablePlugin(SimplePluginManager.java:520) ~[paper-api-1.20.6-R0.1-SNAPSHOT.jar:?]
        at org.bukkit.craftbukkit.CraftServer.enablePlugin(CraftServer.java:626) ~[paper-1.20.6.jar:1.20.6-151-a4f0f5c]
        at org.bukkit.craftbukkit.CraftServer.enablePlugins(CraftServer.java:575) ~[paper-1.20.6.jar:1.20.6-151-a4f0f5c]
        at net.minecraft.server.MinecraftServer.loadWorld0(MinecraftServer.java:675) ~[paper-1.20.6.jar:1.20.6-151-a4f0f5c]
        at net.minecraft.server.MinecraftServer.loadLevel(MinecraftServer.java:437) ~[paper-1.20.6.jar:1.20.6-151-a4f0f5c]
        at net.minecraft.server.dedicated.DedicatedServer.initServer(DedicatedServer.java:323) ~[paper-1.20.6.jar:1.20.6-151-a4f0f5c]
        at net.minecraft.server.MinecraftServer.runServer(MinecraftServer.java:1136) ~[paper-1.20.6.jar:1.20.6-151-a4f0f5c]
        at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:323) ~[paper-1.20.6.jar:1.20.6-151-a4f0f5c]
        at java.base/java.lang.Thread.run(Thread.java:1583) ~[?:?]
Caused by: java.lang.ClassNotFoundException: net.kyori.adventure.text.minimessage.translation.MiniMessageTranslationStore
        at io.papermc.paper.plugin.entrypoint.classloader.PaperPluginClassLoader.loadClass(PaperPluginClassLoader.java:146) ~[paper-1.20.6.jar:1.20.6-151-a4f0f5c]
        at io.papermc.paper.plugin.entrypoint.classloader.PaperPluginClassLoader.loadClass(PaperPluginClassLoader.java:107) ~[paper-1.20.6.jar:1.20.6-151-a4f0f5c]
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:526) ~[?:?]
        ... 13 more
[09:44:21 INFO]: [AntiRedstoneClock-Remastered] Disabling AntiRedstoneClock-Remastered v2.0.13
[09:44:21 ERROR]: Error occurred while disabling AntiRedstoneClock-Remastered v2.0.13
java.lang.NullPointerException: Cannot invoke "net.onelitefeather.antiredstoneclockremastered.service.UpdateService.shutdown()" because "this.updateService" is null
        at AntiRedstoneClock-Remastered-2.0.13.jar/net.onelitefeather.antiredstoneclockremastered.AntiRedstoneClockRemastered.onDisable(AntiRedstoneClockRemastered.java:110) ~[AntiRedstoneClock-Remastered-2.0.13.jar:?]
        at org.bukkit.plugin.java.JavaPlugin.setEnabled(JavaPlugin.java:291) ~[paper-api-1.20.6-R0.1-SNAPSHOT.jar:?]
        at io.papermc.paper.plugin.manager.PaperPluginInstanceManager.disablePlugin(PaperPluginInstanceManager.java:237) ~[paper-1.20.6.jar:1.20.6-151-a4f0f5c]
        at io.papermc.paper.plugin.manager.PaperPluginManagerImpl.disablePlugin(PaperPluginManagerImpl.java:114) ~[paper-1.20.6.jar:1.20.6-151-a4f0f5c]
        at org.bukkit.plugin.SimplePluginManager.disablePlugin(SimplePluginManager.java:550) ~[paper-api-1.20.6-R0.1-SNAPSHOT.jar:?]
        at io.papermc.paper.plugin.manager.PaperPluginInstanceManager.enablePlugin(PaperPluginInstanceManager.java:206) ~[paper-1.20.6.jar:1.20.6-151-a4f0f5c]
        at io.papermc.paper.plugin.manager.PaperPluginManagerImpl.enablePlugin(PaperPluginManagerImpl.java:109) ~[paper-1.20.6.jar:1.20.6-151-a4f0f5c]
        at org.bukkit.plugin.SimplePluginManager.enablePlugin(SimplePluginManager.java:520) ~[paper-api-1.20.6-R0.1-SNAPSHOT.jar:?]
        at org.bukkit.craftbukkit.CraftServer.enablePlugin(CraftServer.java:626) ~[paper-1.20.6.jar:1.20.6-151-a4f0f5c]
        at org.bukkit.craftbukkit.CraftServer.enablePlugins(CraftServer.java:575) ~[paper-1.20.6.jar:1.20.6-151-a4f0f5c]
        at net.minecraft.server.MinecraftServer.loadWorld0(MinecraftServer.java:675) ~[paper-1.20.6.jar:1.20.6-151-a4f0f5c]
        at net.minecraft.server.MinecraftServer.loadLevel(MinecraftServer.java:437) ~[paper-1.20.6.jar:1.20.6-151-a4f0f5c]
        at net.minecraft.server.dedicated.DedicatedServer.initServer(DedicatedServer.java:323) ~[paper-1.20.6.jar:1.20.6-151-a4f0f5c]
        at net.minecraft.server.MinecraftServer.runServer(MinecraftServer.java:1136) ~[paper-1.20.6.jar:1.20.6-151-a4f0f5c]
        at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:323) ~[paper-1.20.6.jar:1.20.6-151-a4f0f5c]
        at java.base/java.lang.Thread.run(Thread.java:1583) ~[?:?]

### Steps to reproduce

Run the server

### Plugin and Datapack List

[10:30:01 INFO]: Server Plugins (42):
[10:30:01 INFO]: Paper Plugins:
[10:30:01 INFO]:  - AntiRedstoneClock-Remastered, FancyNpcs
[10:30:01 INFO]: Bukkit Plugins:
[10:30:01 INFO]:  - AuthMe, BentoBox, BentoBox-Bank, BentoBox-Biomes, BentoBox-BSkyBlock, BentoBox-Challenges, BentoBox-CheckMeOut, BentoBox-Level, BentoBox-Limits, BentoBox-MagicCobblestoneGenerator
[10:30:01 INFO]:  EasyArmorStands, Essentials, EssentialsChat, EssentialsSpawn, FastAsyncWorldEdit, floodgate, Geyser-Spigot, GrimAC, GSit, ImageFrame
[10:30:01 INFO]:  InteractionVisualizer, InvSeePlusPlus, LevelledMobs, LuckPerms, Multiverse-Core, NotQuests, PlaceholderAPI, Plan, ProAntiTab, ProtocolLib
[10:30:01 INFO]:  SilkSpawners_v2, SkinsRestorer, spark, TAB, Vault, ViaBackwards, ViaVersion, voicechat, VoidWorldGenerator, WorldGuard

### Paper version

[10:30:26 INFO]: This server is running Paper version 1.20.6-151-ver/1.20.6@a4f0f5c (2024-10-30T19:13:22Z) (Implementing API version 1.20.6-R0.1-SNAPSHOT)

### Plugin version

[10:33:12 INFO]: AntiRedstoneClock-Remastered version 2.0.13
[10:33:12 INFO]: Authors: OneLiteFeather and TheMeinerLP

### Other

_No response_

--- a/src/main/java/net/onelitefeather/antiredstoneclockremastered/AntiRedstoneClockRemastered.java
+++ b/src/main/java/net/onelitefeather/antiredstoneclockremastered/AntiRedstoneClockRemastered.java
@@ -1,10 +1,8 @@
 package net.onelitefeather.antiredstoneclockremastered;
 
-import net.kyori.adventure.key.Key;
+import io.papermc.paper.ServerBuildInfo;
 import net.kyori.adventure.text.Component;
 import net.kyori.adventure.text.minimessage.MiniMessage;
-import net.kyori.adventure.text.minimessage.translation.MiniMessageTranslationStore;
-import net.kyori.adventure.translation.GlobalTranslator;
 import net.kyori.adventure.util.UTF8ResourceBundleControl;
 import net.onelitefeather.antiredstoneclockremastered.api.PlotsquaredSupport;
 import net.onelitefeather.antiredstoneclockremastered.api.WorldGuardSupport;
@@ -16,6 +14,9 @@ import net.onelitefeather.antiredstoneclockremastered.plotsquared.v6.PlotSquared
 import net.onelitefeather.antiredstoneclockremastered.plotsquared.v7.PlotSquaredModernSupport;
 import net.onelitefeather.antiredstoneclockremastered.service.RedstoneClockService;
 import net.onelitefeather.antiredstoneclockremastered.service.UpdateService;
+import net.onelitefeather.antiredstoneclockremastered.service.api.TranslationService;
+import net.onelitefeather.antiredstoneclockremastered.service.impl.LegacyTranslationService;
+import net.onelitefeather.antiredstoneclockremastered.service.impl.ModernTranslationService;
 import net.onelitefeather.antiredstoneclockremastered.utils.CheckTPS;
 import net.onelitefeather.antiredstoneclockremastered.worldguard.v6.WorldGuardLegacySupport;
 import net.onelitefeather.antiredstoneclockremastered.worldguard.v7.WorldGuardModernSupport;
@@ -44,6 +45,7 @@ import java.util.*;
 import static org.incendo.cloud.parser.standard.StringParser.greedyStringParser;
 
 public final class AntiRedstoneClockRemastered extends JavaPlugin {
+    public static final String RESOURCE_BUNDLE_NAME = "antiredstoneclockremasterd";
     private CheckTPS tps;
 
     private RedstoneClockService redstoneClockService;
@@ -67,27 +69,30 @@ public final class AntiRedstoneClockRemastered extends JavaPlugin {
 
     @Override
     public void onEnable() {
-        final MiniMessageTranslationStore miniMessageTranslationStore = MiniMessageTranslationStore.create(Key.key("antiredstoneclockremastered", "translations"));
-        miniMessageTranslationStore.defaultLocale(Locale.US);
+        TranslationService translationService;
+        ServerBuildInfo buildInfo = ServerBuildInfo.buildInfo();
+        if (buildInfo.minecraftVersionId().startsWith("1.20")) {
+            translationService = new LegacyTranslationService();
+            getSLF4JLogger().info("Using legacy translation service");
+        } else {
+            translationService = new ModernTranslationService();
+            getSLF4JLogger().info("Using modern translation service");
+        }
         Path langFolder = getDataFolder().toPath().resolve("lang");
+        if (Files.notExists(langFolder)) {
+            try {
+                Files.createDirectories(langFolder);
+            } catch (IOException e) {
+                getSLF4JLogger().error("An error occurred while creating lang folder");
+                return;
+            }
+        }
         var languages = new HashSet<>(getConfig().getStringList("translations"));
         languages.add("en-US");
-        if (Files.exists(langFolder)) {
-            try (var urlClassLoader = new URLClassLoader(new URL[]{langFolder.toUri().toURL()})) {
-                languages.stream().map(Locale::forLanguageTag).forEach(r -> {
-                    var bundle = ResourceBundle.getBundle("antiredstoneclockremasterd", r, urlClassLoader, UTF8ResourceBundleControl.get());
-                    miniMessageTranslationStore.registerAll(r, bundle, false);
-                });
-            } catch (IOException e) {
-                throw new RuntimeException(e);
-            }
-        } else {
-            languages.stream().map(Locale::forLanguageTag).forEach(r -> {
-                var bundle = ResourceBundle.getBundle("antiredstoneclockremasterd", r, UTF8ResourceBundleControl.get());
-                miniMessageTranslationStore.registerAll(r, bundle, false);
-            });
-        }
-        GlobalTranslator.translator().addSource(miniMessageTranslationStore);
+        languages.stream()
+            .map(Locale::forLanguageTag)
+            .forEach(locale -> loadAndRegisterTranslation(locale, langFolder, translationService));
+        translationService.registerGlobal();
         donationInformation();
         updateService();
         enableCommandFramework();
@@ -107,7 +112,9 @@ public final class AntiRedstoneClockRemastered extends JavaPlugin {
 
     @Override
     public void onDisable() {
-        this.updateService.shutdown();
+        if (this.updateService != null) {
+            this.updateService.shutdown();
+        }
     }
 
     private void donationInformation() {
@@ -274,6 +281,29 @@ public final class AntiRedstoneClockRemastered extends JavaPlugin {
             return this.worldGuardSupport.getVersion();
         }
         return "unknown";
+    }
+
+    private void loadAndRegisterTranslation(Locale locale, Path langFolder, TranslationService translationService) {
+        try {
+            ResourceBundle bundle = loadResourceBundle(locale, langFolder);
+            if (bundle != null) {
+                translationService.registerAll(locale, bundle, false);
+            }
+        } catch (Exception e) {
+            getSLF4JLogger().error("An error occurred while loading language file for locale {}", locale, e);
+        }
+    }
+
+    private ResourceBundle loadResourceBundle(Locale locale, Path langFolder) throws Exception {
+        Path langFile = langFolder.resolve(RESOURCE_BUNDLE_NAME + "_" + locale.toLanguageTag() + ".properties");
+
+        if (Files.exists(langFile)) {
+            try (URLClassLoader urlClassLoader = new URLClassLoader(new URL[]{langFolder.toUri().toURL()})) {
+                return ResourceBundle.getBundle(RESOURCE_BUNDLE_NAME, locale, urlClassLoader, UTF8ResourceBundleControl.get());
+            }
+        } else {
+            return ResourceBundle.getBundle(RESOURCE_BUNDLE_NAME, locale, UTF8ResourceBundleControl.get());
+        }
     }
 
     public CheckTPS getTps() {

--- a/src/main/java/net/onelitefeather/antiredstoneclockremastered/service/api/TranslationService.java
+++ b/src/main/java/net/onelitefeather/antiredstoneclockremastered/service/api/TranslationService.java
@@ -1,0 +1,14 @@
+package net.onelitefeather.antiredstoneclockremastered.service.api;
+
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Locale;
+import java.util.ResourceBundle;
+
+public interface TranslationService {
+
+    void registerAll(final @NotNull Locale locale, final @NotNull ResourceBundle bundle, final boolean escapeSingleQuotes);
+
+    void registerGlobal();
+
+}

--- a/src/main/java/net/onelitefeather/antiredstoneclockremastered/service/impl/LegacyTranslationService.java
+++ b/src/main/java/net/onelitefeather/antiredstoneclockremastered/service/impl/LegacyTranslationService.java
@@ -1,0 +1,32 @@
+package net.onelitefeather.antiredstoneclockremastered.service.impl;
+
+import net.kyori.adventure.key.Key;
+import net.kyori.adventure.translation.GlobalTranslator;
+import net.kyori.adventure.translation.TranslationRegistry;
+import net.onelitefeather.antiredstoneclockremastered.service.api.TranslationService;
+import net.onelitefeather.antiredstoneclockremastered.translations.PluginTranslationRegistry;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Locale;
+import java.util.ResourceBundle;
+
+@Deprecated(since = "2.0.13", forRemoval = true)
+public final class LegacyTranslationService implements TranslationService {
+
+    private final TranslationRegistry translationRegistry;
+
+    public LegacyTranslationService() {
+        this.translationRegistry = new PluginTranslationRegistry(TranslationRegistry.create(Key.key("antiredstoneclockremastered", "translations")));
+        this.translationRegistry.defaultLocale(Locale.US);
+    }
+
+    @Override
+    public void registerAll(@NotNull Locale locale, @NotNull ResourceBundle bundle, boolean escapeSingleQuotes) {
+        this.translationRegistry.registerAll(locale, bundle, escapeSingleQuotes);
+    }
+
+    @Override
+    public void registerGlobal() {
+        GlobalTranslator.translator().addSource(this.translationRegistry);
+    }
+}

--- a/src/main/java/net/onelitefeather/antiredstoneclockremastered/service/impl/ModernTranslationService.java
+++ b/src/main/java/net/onelitefeather/antiredstoneclockremastered/service/impl/ModernTranslationService.java
@@ -1,0 +1,30 @@
+package net.onelitefeather.antiredstoneclockremastered.service.impl;
+
+import net.kyori.adventure.key.Key;
+import net.kyori.adventure.text.minimessage.translation.MiniMessageTranslationStore;
+import net.kyori.adventure.translation.GlobalTranslator;
+import net.onelitefeather.antiredstoneclockremastered.service.api.TranslationService;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Locale;
+import java.util.ResourceBundle;
+
+public final class ModernTranslationService implements TranslationService {
+
+    private final MiniMessageTranslationStore miniMessageTranslationStore;
+
+    public ModernTranslationService() {
+        this.miniMessageTranslationStore = MiniMessageTranslationStore.create(Key.key("antiredstoneclockremastered", "translations"));
+        this.miniMessageTranslationStore.defaultLocale(Locale.US);
+    }
+
+    @Override
+    public void registerAll(@NotNull Locale locale, @NotNull ResourceBundle bundle, boolean escapeSingleQuotes) {
+        this.miniMessageTranslationStore.registerAll(locale, bundle, escapeSingleQuotes);
+    }
+
+    @Override
+    public void registerGlobal() {
+        GlobalTranslator.translator().addSource(this.miniMessageTranslationStore);
+    }
+}

--- a/src/main/java/net/onelitefeather/antiredstoneclockremastered/translations/PluginTranslationRegistry.java
+++ b/src/main/java/net/onelitefeather/antiredstoneclockremastered/translations/PluginTranslationRegistry.java
@@ -1,0 +1,125 @@
+package net.onelitefeather.antiredstoneclockremastered.translations;
+
+import net.kyori.adventure.key.Key;
+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.text.ComponentLike;
+import net.kyori.adventure.text.TranslatableComponent;
+import net.kyori.adventure.text.minimessage.Context;
+import net.kyori.adventure.text.minimessage.MiniMessage;
+import net.kyori.adventure.text.minimessage.ParsingException;
+import net.kyori.adventure.text.minimessage.tag.Tag;
+import net.kyori.adventure.text.minimessage.tag.resolver.ArgumentQueue;
+import net.kyori.adventure.text.minimessage.tag.resolver.TagResolver;
+import net.kyori.adventure.translation.TranslationRegistry;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.text.MessageFormat;
+import java.util.List;
+import java.util.Locale;
+import java.util.Objects;
+
+public final class PluginTranslationRegistry implements TranslationRegistry {
+
+    private final TranslationRegistry backedRegistry;
+
+    public PluginTranslationRegistry(TranslationRegistry backedRegistry) {
+        this.backedRegistry = backedRegistry;
+    }
+
+    @Override
+    public boolean contains(@NotNull String key) {
+        return backedRegistry.contains(key);
+    }
+
+    @Override
+    public boolean contains(@NotNull String key, @NotNull Locale locale) {
+        return backedRegistry.contains(key, locale);
+    }
+
+    @Override
+    public @NotNull Key name() {
+        return backedRegistry.name();
+    }
+
+    @Override
+    public @Nullable MessageFormat translate(@NotNull String key, @NotNull Locale locale) {
+        return null;
+    }
+
+    @Override
+    public @Nullable Component translate(
+            @NotNull TranslatableComponent component,
+            @NotNull Locale locale
+    ) {
+        final MessageFormat translationFormat = backedRegistry.translate(component.key(), locale);
+
+        if (translationFormat == null) {
+            return null;
+        }
+
+        final String miniMessageString = translationFormat.toPattern();
+
+        final Component resultingComponent;
+
+        if (component.arguments().isEmpty()) {
+            resultingComponent = MiniMessage.miniMessage().deserialize(miniMessageString);
+        } else {
+            resultingComponent = MiniMessage.miniMessage().deserialize(miniMessageString,
+                    new ArgumentTag(component.arguments()));
+        }
+
+        if (component.children().isEmpty()) {
+            return resultingComponent;
+        } else {
+            return resultingComponent.children(component.children());
+        }
+    }
+
+    @Override
+    public void defaultLocale(@NotNull Locale locale) {
+        backedRegistry.defaultLocale(locale);
+    }
+
+    @Override
+    public void register(@NotNull String key, @NotNull Locale locale, @NotNull MessageFormat format) {
+        backedRegistry.register(key, locale, format);
+    }
+
+    @Override
+    public void unregister(@NotNull String key) {
+        backedRegistry.unregister(key);
+    }
+
+    private static final class ArgumentTag implements TagResolver {
+        private static final String NAME = "argument";
+        private static final String NAME_1 = "arg";
+
+        private final List<? extends ComponentLike> argumentComponents;
+
+        private ArgumentTag(final @NotNull  List<? extends ComponentLike> argumentComponents) {
+            this.argumentComponents = Objects.requireNonNull(argumentComponents, "argumentComponents");
+        }
+
+        @Override
+        public Tag resolve(
+                final @NotNull String name,
+                final @NotNull ArgumentQueue arguments,
+                final @NotNull Context ctx
+        ) throws ParsingException {
+            final int index = arguments.popOr("No argument number provided")
+                    .asInt().orElseThrow(() -> ctx.newException("Invalid argument number", arguments));
+
+            if (index < 0 || index >= argumentComponents.size()) {
+                throw ctx.newException("Invalid argument number", arguments);
+            }
+
+            return Tag.inserting(argumentComponents.get(index));
+        }
+
+        @Override
+        public boolean has(final @NotNull String name) {
+            return name.equals(NAME) || name.equals(NAME_1);
+        }
+    }
+}

